flow_transforms.py

```diff
@@ -182,14 +182,22 @@ def __call__(self, inputs,target):
         angle1 = applied_angle - diff/2
         angle2 = applied_angle + diff/2
         angle1_rad = angle1*np.pi/180
+        diff_rad = diff*np.pi/180
 
         h, w, _ = target.shape
 
-        def rotate_flow(i,j,k):
-            return -k*(j-w/2)*(diff*np.pi/180) + (1-k)*(i-h/2)*(diff*np.pi/180)
+        warped_coords = np.mgrid[:w, :h].T + target
```

Comment:
Is  `np.mgrid` a global variable, or do we need to recreate it each time ? Since we call it with [] instead of (), i would say yes, but I haven't found much info about this

Reply:
Well, it's an "index trick" function from numpy. You use the bracket notation but it really is a function call. https://numpy.org/doc/stable/reference/generated/numpy.mgrid.html#numpy.mgrid

We could potentially store the grid for subsequent calls, but it would be invalidated if the image size changes.

Reply:
I think it would not be invalidated, no matter the size of the array, if `i <= h` and `j<= w` then we get mgrid[i,j] = [i,j]
The problem comes up if you start with small images and then follow up with bigger images. I tried to solve the problem in my other repo, see here : https://github.com/ClementPinard/SfmLearner-Pytorch/blob/master/inverse_warp.py#L36

Reply:
Totally agree. I'm not sure if the speed gain vs. increased code complexity is really worth it though.

Reply:
Since this part is multi threaded during the data loading (contrary to the `inverse_warp` helper I linked), I guess it shouldn't be too much of a problem. What is your hardware ? Do you succeed in making the [batch time](https://github.com/ClementPinard/FlowNetPytorch/blob/master/main.py#L263) zero with enough jobs running in parallel ?

Reply:
Since I can't plug an SSD on my server, my training is mostly limited by disk accesses. I've tried to circumvent this by converting databases to [Webdataset](https://github.com/tmbdev/webdataset).
In case of a CPU-limited training, something worth trying is to fetch data from disk in parallel then perform augmentation in the main process on the GPU. For example by converting data augmentations classes to torch.nn.module (+ using torch.jit.script).

Reply:
Indeed, it would work.
The goal of this repo is to be a simple basis to work on to propose new optical flow architecture, I'll leave it as is. I'll try to run test on my own to see if the optimization we discussed is actually worth the code complexification ;)